circular-buffer/src/lib.rs

```diff
@@ -0,0 +1,818 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+#![warn(missing_docs)]
+
+//! # `circular-buffer`
+//!
+//! An optimized FIFO queue that allows safe access to the internal storage as a slice (i.e. not
+//! just element-by-element). This is useful for circular buffers of bytes. Since it uses
+//! `smallvec`'s `Array` trait it can only be backed by an array of static size, this may change in
+//! the future.
+
+extern crate smallvec;
+
+use std::ops::Drop;
+use std::mem::ManuallyDrop;
+
+use smallvec::Array;
+
+use owned_slice::OwnedSlice;
+
+/// A slice that owns its elements, but not their storage. This is useful for things like
+/// `Vec::retain` and `CircularBuffer::pop_slice`, since these functions can return a slice but the
+/// elements of these slices would be leaked after the slice goes out of scope. `OwnedSlice` simply
+/// manually drops all its elements when it goes out of scope.
+pub mod owned_slice {
+    use std::ops::{Deref, DerefMut, Drop};
+    use std::mem::ManuallyDrop;
+
+    /// A slice that owns its elements, but not their storage. This is useful for things like
+    /// `Vec::retain` and `CircularBuffer::pop_slice`, since these functions can return a slice but
+    /// the elements of these slices would be leaked after the slice goes out of scope. `OwnedSlice`
+    /// simply manually drops all its elements when it goes out of scope.
+    #[derive(Debug, Eq, PartialEq)]
+    pub struct OwnedSlice<'a, T: 'a>(&'a mut [T]);
+
+    /// Owning iterator for `OwnedSlice`.
+    pub struct IntoIter<'a, T: 'a> {
+        slice: ManuallyDrop<OwnedSlice<'a, T>>,
```

Comment:
`OwnedSlice`'s destructor drops all elements in place, even those which have already been read out by the iterator

Reply:
(can I just say it again: please don't use any unsafe code at all!)